packages/parallel/init.lua

```diff
@@ -99,75 +438,196 @@ function package:_init (options)
          end
-         SU.debug(
-            "parallel",
-            frame,
-            ": pre-sync content=",
-            calculations[frame].mark,
-            ", now",
-            #typesetter.state.outputQueue,
-            ", height of material:",
-            calculations[frame].heightOfNewMaterial
-         )
       end)
+
+      -- Add balancing glue to align frame heights.
       addBalancingGlue(maxheight)
+
+      -- Handle parskip (spacing between successive frames).
+      local parskip = SILE.settings:get("document.parskip")
+
+      if not parskip.length then
+         -- Insert default parskip value if not defined.
+         addParskipToFrames(SILE.types.length("12pt plus 3pt minus 1pt"))
+      else
+         -- Use user-defined parskip value.
+         addParskipToFrames(parskip)
+      end
+   end)
+
+   self:registerCommand("smaller", function(_, content)
+      SILE.settings:temporarily(function()
+         local currentSize = SILE.settings:get("font.size")
+         SILE.settings:set("font.size", currentSize * 0.75) -- Scale down to 75%
+         SILE.settings:set("font.weight", 800)
+         SILE.process(content)
+      end)
+   end)
+
+   self:registerCommand("footnoteNumber", function(options, content)
+      local height = options.height or "0.3em" -- Default height for superscripts
+      SILE.call("raise", { height = height }, function()
+         SILE.call("smaller", {}, function()
+            SILE.process(content)
+         end)
+      end)
+   end)
+
+   -- Adapted from the `resilient.footnotes` package.
+   -- Defines `parallel_footnote:rule`, a helper command for setting a footnote rule.
+   self:registerCommand("parallel_footnote:rule", function(options, _)
+      local width = SU.cast("measurement", options.width or "20%fw") -- Default: 1/5 of the text block width.
+      local beforeskipamount = SU.cast("vglue", options.beforeskipamount or "1ex") -- Space before the rule.
+      local afterskipamount = SU.cast("vglue", options.afterskipamount or "1ex") -- Space after the rule.
+      local thickness = SU.cast("measurement", options.thickness or "0.5pt") -- Thickness of the rule.
+
+      SILE.call("noindent")
+      SILE.call("rebox", {}, function()
+         SILE.call("hrule", { width = width, height = thickness })
+      end)
+      SILE.typesetter:leaveHmode()
+      SILE.typesetter:pushExplicitVglue(afterskipamount)
+   end, "Helper command for setting a footnote rule.")
+
+   self:registerCommand("parallel_footnote:constructor", function(options, content)
+      local markerText = options.marker or "?" -- Default marker if none provided
+
+      SILE.settings:temporarily(function()
+         -- Set font footnotes, 80% of current font size
+         SILE.settings:set("font.size", SILE.settings:get("font.size") * 0.80)
+
+         -- Measure the marker width
+         local markerWidth = measureStringWidth(markerText)
```

Comment:
You are computing the width by shaping the marker text. This will be shaped _again_ below when you typeset it (`SILE.typesetter:typeset(markerText)`). I'd suggest rather using a box: you'd get the width and can just re-use it as-is afterwards, without re-computing costs under the hood.
```
local markerBox = SILE:typesetter:makeHbox({ markerText })
local markerWidth = markerBox.width
...
SILE.typesetter:pushHbox(markBox) -- don't re-typeset, just push our box

Reply:
> Looked into your resilent.footnotes but did not find it,

Hehe.... The logic is entirely abstracted and buried  deep into the `style:apply:number` logic (in the resilient.styles package), 'cause every styling decision is customizable in _re·sil·ient_, and the footnote package doesn't have to care, it just applies a style where needed.